Classes/Controller/GetFileController.php

```diff
@@ -95,48 +99,74 @@ public function attachmentAction()
 
                     header('Content-Type: text/xml; charset=UTF-8');
 
-                }
+                    return $metsXml;
+
+                } else {
+
+                    http_response_code(404);
 
-                return $metsXml;
+                    print('No such document');
+
+                    exit;
+
+                }
 
             case 'attachment':
                 $path = rtrim('http://' . $fedoraHost, "/") . '/fedora/objects/' . $piVars['qid'] . '/datastreams/' . $piVars['attachment'] . '/content';
                 break;
+
             default:
-                break;
+                http_response_code(404);
+                print('No such action');
+                exit;
         }
 
         // get remote header and set it before passtrough
         $headers = get_headers($path);
 
-        foreach ($headers as $key => $value) {
-            // set remote header information
-            preg_match('/filename="(.*)"/', $value, $fileName);
+        if (FALSE === $headers) {
+            http_response_code(500);
+            print('Error while fetching headers');
+            exit;
+        }
+
+        foreach ($headers as $value) {
 
-            if ($fileName[1]) {
-                header('Content-Disposition: inline; filename="' . $fileName[1] . '";');
+            if (FALSE !== stripos($value, 'Content-Disposition')) {
+                header($value);
                 continue;
             }
 
-            if (substr($value, 0, 13) == "Content-Type:") {
+            if (FALSE !== stripos($value, 'Content-Type')) {
                 header($value);
                 continue;
             }
 
-            if (substr($value, 0, 13) == "Content-Length:") {
+            if (FALSE !== stripos($value, 'Content-Length')) {
                 header($value);
                 continue;
             }
         }
 
         if ($stream = fopen($path, 'r')) {
 
+            // close active session if any
+            session_write_close();
+
+            // stop output buffering
+            ob_end_clean();
+
             fpassthru($stream);
 
             fclose($stream);
 
```

Comment:
Her i'd work with `exit` to quit right after the file streaming is done. There is no use-case where anything should follow the stream.